File: fraiseql-php/tests/RoleBasedAccessControlTest.php
<?php

declare(strict_types=1);

namespace FraiseQL\Tests;

use PHPUnit\Framework\TestCase;
use FraiseQL\SchemaRegistry;
use FraiseQL\Security\RoleRequiredBuilder;
use FraiseQL\Security\RoleMatchStrategy;
use FraiseQL\Attributes\RoleRequired;

/**
 * Tests for role-based access control.
 */
final class RoleBasedAccessControlTest extends TestCase
{
    protected function tearDown(): void
    {
        SchemaRegistry::getInstance()->clear();
        parent::tearDown();
    }

    public function testRoleRequiredSingleRole(): void
    {
        $config = RoleRequiredBuilder::create()
            ->roles('admin')
            ->description('Admin role required')
            ->build();

        $this->assertCount(1, $config->roles);
        $this->assertContains('admin', $config->roles);
    }

    public function testRoleRequiredMultipleRoles(): void
    {
        $config = RoleRequiredBuilder::create()
            ->roles('manager', 'director')
            ->description('Manager or director required')
            ->build();

        $this->assertCount(2, $config->roles);
        $this->assertContains('manager', $config->roles);
        $this->assertContains('director', $config->roles);
    }

    public function testRoleRequiredRolesArray(): void
    {
        $config = RoleRequiredBuilder::create()
            ->rolesArray(['viewer', 'editor', 'admin'])
            ->description('Multiple roles via array')
            ->build();

        $this->assertCount(3, $config->roles);
    }

    public function testRoleMatchStrategyAny(): void
    {
        $config = RoleRequiredBuilder::create()
            ->roles('manager', 'director')
            ->strategy(RoleMatchStrategy::ANY)
            ->description('User needs at least one role')
            ->build();

        $this->assertSame(RoleMatchStrategy::ANY, $config->strategy);
    }

    public function testRoleMatchStrategyAll(): void
    {
        $config = RoleRequiredBuilder::create()
            ->roles('admin', 'auditor')
            ->strategy(RoleMatchStrategy::ALL)
            ->description('User needs all roles')
            ->build();

        $this->assertSame(RoleMatchStrategy::ALL, $config->strategy);
    }

    public function testRoleMatchStrategyExactly(): void
    {
        $config = RoleRequiredBuilder::create()
            ->roles('admin')
            ->strategy(RoleMatchStrategy::EXACTLY)
            ->description('User must have exactly these roles')
            ->build();

        $this->assertSame(RoleMatchStrategy::EXACTLY, $config->strategy);
    }

    public function testRoleHierarchy(): void
    {
        $config = RoleRequiredBuilder::create()
            ->roles('user')
            ->hierarchy(true)
            ->description('Role hierarchy enabled')
            ->build();

        $this->assertTrue($config->hierarchy);
    }

    public function testRoleInheritance(): void
    {
        $config = RoleRequiredBuilder::create()
            ->roles('editor')
            ->inherit(true)
            ->description('Inherit role requirements')
            ->build();

        $this->assertTrue($config->inherit);
    }

    public function testRoleOperationSpecific(): void
    {
        $config = RoleRequiredBuilder::create()
            ->roles('admin')
            ->operations('delete,create')
            ->description('Admin for destructive operations')
            ->build();

        $this->assertSame('delete,create', $config->operations);
    }

    public function testRoleCaching(): void
    {
        $config = RoleRequiredBuilder::create()
            ->roles('viewer')
            ->cacheable(true)
            ->cacheDurationSeconds(1800)
            ->build();

        $this->assertTrue($config->cacheable);
        $this->assertSame(1800, $config->cacheDurationSeconds);
    }

    public function testRoleErrorMessage(): void
    {
        $config = RoleRequiredBuilder::create()
            ->roles('admin')
            ->errorMessage('You must be an administrator to access this resource')
            ->build();

        $this->assertSame('You must be an administrator to access this resource', $config->errorMessage);
    }

    public function testRoleFluentChaining(): void
    {
        $config = RoleRequiredBuilder::create()
            ->roles('manager', 'director')
            ->strategy(RoleMatchStrategy::ANY)
            ->hierarchy(true)
            ->description('Manager or director with hierarchy')
            ->errorMessage('Insufficient role')
            ->operations('read,update')
            ->inherit(false)
            ->cacheable(true)
            ->cacheDurationSeconds(900)
            ->build();

        $this->assertCount(2, $config->roles);
        $this->assertSame(RoleMatchStrategy::ANY, $config->strategy);
        $this->assertTrue($config->hierarchy);
        $this->assertFalse($config->inherit);
        $this->assertSame(900, $config->cacheDurationSeconds);
    }

    public function testRoleAdminPattern(): void
    {
        $config = RoleRequiredBuilder::create()
            ->roles('admin')
            ->strategy(RoleMatchStrategy::EXACTLY)
            ->hierarchy(true)
            ->description('Full admin access with hierarchy')
            ->build();

        $this->assertCount(1, $config->roles);
        $this->assertTrue($config->hierarchy);
    }

    public function testRoleManagerPattern(): void
    {
        $config = RoleRequiredBuilder::create()
            ->roles('manager', 'director', 'executive')
            ->strategy(RoleMatchStrategy::ANY)
            ->description('Management tier access')
            ->operations('read,create,update')
            ->build();

        $this->assertCount(3, $config->roles);
        $this->assertSame('read,create,update', $config->operations);
    }

    public function testRoleDataScientistPattern(): void
    {
        $config = RoleRequiredBuilder::create()
            ->roles('data_scientist', 'analyst')
            ->strategy(RoleMatchStrategy::ANY)
            ->description('Data access for scientists and analysts')
            ->operations('read')
            ->build();

        $this->assertCount(2, $config->roles);
    }

    public function testRoleAttributeBasic(): void
    {
        #[RoleRequired(roles: ['admin'])]
        class AdminPanel
        {
        }

        $this->assertTrue(true);
    }

    public function testRoleAttributeWithStrategy(): void
    {
        #[RoleRequired(
            roles: ['manager', 'director'],
            strategy: RoleMatchStrategy::ANY,
            description: 'Management access',
        )]
        class SalaryData
        {
        }

        $this->assertTrue(true);
    }

    public function testRoleAttributeAllParameters(): void
    {
        #[RoleRequired(
            roles: ['admin', 'auditor'],
            strategy: RoleMatchStrategy::ALL,
            hierarchy: true,
            description: 'Full admin with auditor',
            errorMessage: 'Insufficient privileges',
            operations: 'delete,create',
            inherit: false,
            cacheable: true,
            cacheDurationSeconds: 1200,
        )]
        class ComplexRoleRequirement
        {
        }

        $this->assertTrue(true);
    }

    public function testMultipleRolesWithDifferentStrategies(): void
    {
        $any = RoleRequiredBuilder::create()
            ->roles('editor', 'contributor')
            ->strategy(RoleMatchStrategy::ANY)
            ->build();

        $all = RoleRequiredBuilder::create()
            ->roles('editor', 'reviewer')
            ->strategy(RoleMatchStrategy::ALL)
            ->build();

        $exactly = RoleRequiredBuilder::create()
            ->roles('admin')
            ->strategy(RoleMatchStrategy::EXACTLY)
            ->build();

        $this->assertSame(RoleMatchStrategy::ANY, $any->strategy);
        $this->assertSame(RoleMatchStrategy::ALL, $all->strategy);
        $this->assertSame(RoleMatchStrategy::EXACTLY, $exactly->strategy);
    }

    public function testRoleHierarchyPatterns(): void
    {
        // Admin > Manager > Employee hierarchy
        $admin = RoleRequiredBuilder::create()
            ->roles('admin')
            ->hierarchy(true)
            ->build();

        $manager = RoleRequiredBuilder::create()
            ->roles('manager')
            ->hierarchy(true)
            ->build();

        $employee = RoleRequiredBuilder::create()
            ->roles('employee')
            ->hierarchy(false)
            ->build();

        $this->assertTrue($admin->hierarchy);
        $this->assertTrue($manager->hierarchy);
        $this->assertFalse($employee->hierarchy);
    }
}

File: fraiseql-csharp/tests/FraiseQL.Security.Tests/RoleBasedAccessControlTest.cs
using Xunit;
using FraiseQL.Security;
using System.Collections.Generic;
using System.Linq;

namespace FraiseQL.Security.Tests
{
    public class RoleBasedAccessControlTest
    {
        [Fact]
        public void ShouldCreateSingleRoleRequirement()
        {
            var config = new RoleRequiredBuilder()
                .Roles("admin")
                .Build();

            Assert.Single(config.Roles);
            Assert.Equal("admin", config.Roles[0]);
        }

        [Fact]
        public void ShouldCreateMultipleRoleRequirements()
        {
            var config = new RoleRequiredBuilder()
                .Roles("manager", "director")
                .Build();

            Assert.Equal(2, config.Roles.Count);
            Assert.Contains("manager", config.Roles);
            Assert.Contains("director", config.Roles);
        }

        [Fact]
        public void ShouldUseAnyRoleMatchingStrategy()
        {
            var config = new RoleRequiredBuilder()
                .Roles("viewer", "editor")
                .Strategy(new RoleMatchStrategy.Any())
                .Description("At least one role")
                .Build();

            Assert.IsType<RoleMatchStrategy.Any>(config.Strategy);
            Assert.Equal("any", config.Strategy.Value);
        }

        [Fact]
        public void ShouldUseAllRoleMatchingStrategy()
        {
            var config = new RoleRequiredBuilder()
                .Roles("admin", "auditor")
                .Strategy(new RoleMatchStrategy.All())
                .Description("All roles required")
                .Build();

            Assert.IsType<RoleMatchStrategy.All>(config.Strategy);
            Assert.Equal("all", config.Strategy.Value);
        }

        [Fact]
        public void ShouldUseExactlyRoleMatchingStrategy()
        {
            var config = new RoleRequiredBuilder()
                .Roles("exact_role")
                .Strategy(new RoleMatchStrategy.Exactly())
                .Description("Exactly these roles")
                .Build();

            Assert.IsType<RoleMatchStrategy.Exactly>(config.Strategy);
            Assert.Equal("exactly", config.Strategy.Value);
        }

        [Fact]
        public void ShouldSupportRoleHierarchy()
        {
            var config = new RoleRequiredBuilder()
                .Roles("admin")
                .Hierarchy(true)
                .Description("With hierarchy")
                .Build();

            Assert.True(config.Hierarchy);
        }

        [Fact]
        public void ShouldSupportRoleInheritance()
        {
            var config = new RoleRequiredBuilder()
                .Roles("editor")
                .Inherit(true)
                .Description("Inherits from parent")
                .Build();

            Assert.True(config.Inherit);
        }

        [Fact]
        public void ShouldSetOperationSpecificRoles()
        {
            var config = new RoleRequiredBuilder()
                .Roles("editor")
                .Operations("create,update")
                .Description("Only for edit operations")
                .Build();

            Assert.Equal("create,update", config.Operations);
        }

        [Fact]
        public void ShouldSetCustomErrorMessage()
        {
            var config = new RoleRequiredBuilder()
                .Roles("admin")
                .ErrorMessage("Administrator access required")
                .Build();

            Assert.Equal("Administrator access required", config.ErrorMessage);
        }

        [Fact]
        public void ShouldConfigureCaching()
        {
            var config = new RoleRequiredBuilder()
                .Roles("viewer")
                .Cacheable(true)
                .CacheDurationSeconds(1800)
                .Build();

            Assert.True(config.Cacheable);
            Assert.Equal(1800, config.CacheDurationSeconds);
        }

        [Fact]
        public void ShouldCreateAdminPattern()
        {
            var config = new RoleRequiredBuilder()
                .Roles("admin")
                .Strategy(new RoleMatchStrategy.Any())
                .Description("Admin access")
                .Build();

            Assert.Single(config.Roles);
            Assert.Equal("admin", config.Roles[0]);
        }

        [Fact]
        public void ShouldCreateManagerDirectorPattern()
        {
            var config = new RoleRequiredBuilder()
                .Roles("manager", "director")
                .Strategy(new RoleMatchStrategy.Any())
                .Description("Managers and directors")
                .Build();

            Assert.Equal(2, config.Roles.Count);
            Assert.Equal("any", config.Strategy.Value);
        }

        [Fact]
        public void ShouldCreateDataScientistPattern()
        {
            var config = new RoleRequiredBuilder()
                .Roles("data_scientist", "analyst")
                .Strategy(new RoleMatchStrategy.Any())
                .Description("Data professionals")
                .Build();

            Assert.Equal(2, config.Roles.Count);
        }

        [Fact]
        public void ShouldConvertToDict()
        {
            var config = new RoleRequiredBuilder()
                .Roles("admin", "editor")
                .Strategy(new RoleMatchStrategy.Any())
                .Build();

            var dict = config.ToDict();

            Assert.NotNull(dict);
            Assert.NotNull(dict["roles"]);
            Assert.Equal("any", dict["strategy"]);
        }

        [Fact]
        public void ShouldUseAnnotationAttribute()
        {
            var type = typeof(SalaryData);
            var attr = type.GetCustomAttributes(typeof(RoleRequiredAttribute), false)
                .FirstOrDefault() as RoleRequiredAttribute;

            Assert.NotNull(attr);
            Assert.Contains("manager", attr.Roles);
            Assert.Contains("director", attr.Roles);
        }

        [Fact]
        public void ShouldSetDescription()
        {
            var config = new RoleRequiredBuilder()
                .Roles("viewer")
                .Description("Read-only access")
                .Build();

            Assert.Equal("Read-only access", config.Description);
        }

        [Fact]
        public void ShouldReturnDefaultValues()
        {
            var config = new RoleRequiredBuilder()
                .Roles("user")
                .Build();

            Assert.False(config.Hierarchy);
            Assert.False(config.Inherit);
            Assert.True(config.Cacheable);
            Assert.Equal(300, config.CacheDurationSeconds);
        }
    }

    [RoleRequired(Roles = new[] { "manager", "director" }, Strategy = "any")]
    public class SalaryData
    {
        public string EmployeeId { get; set; } = "";
        public double Salary { get; set; }
    }
}
